R/cache_datastore.r

```diff
@@ -0,0 +1,125 @@
+#' datastore_cache
+#'
+#' Use google datastore to store and retrieve cache items. Requires authentication.
+#'
+#' @seealso \url{https://cloud.google.com/}
+#' @seealso \url{https://cloud.google.com/datastore/docs/concepts/overview}
+#' @seealso \url{https://developers.google.com/identity/protocols/OAuth2#basicsteps}
+#'
+#' @export
+
+cache_datastore <- function(project, cache_name = "rcache") {
+
+  options("googleAuthR.scopes.selected" = c("https://www.googleapis.com/auth/datastore",
+                                            "https://www.googleapis.com/auth/userinfo.email"))
+
+  googleAuthR::gar_auth()
+
+  base_url <- paste0("https://datastore.googleapis.com/v1beta3/projects/", project)
+
+  transaction <- googleAuthR::gar_api_generator(paste0(base_url, ":beginTransaction"),
+                    "POST",
+                    data_parse_function = function(x) x$transaction)
+
+  commit_ds <- googleAuthR::gar_api_generator(paste0(base_url, ":commit"),
+                                            "POST",
+                                            data_parse_function = function(x) x)
+
+  load_ds <- googleAuthR::gar_api_generator(paste0(base_url, ":lookup"),
+                                            "POST",
+                                            data_parse_function = function(resp) {
+                                              # Unserialize and return
+                                              if ("found" %in% names(resp)) {
+                                                resp <- resp$found
+                                                value <- resp$entity$properties$object$blobValue
+                                                response <- unserialize(memDecompress(base64enc::base64decode(value), type = "gzip"))
+                                              } else if ("missing" %in% names(resp)) {
+                                                "!cache-not-found"
+                                              } else {
+                                                stop("Error")
+                                              }
+                                            })
+
+  query_ds <- googleAuthR::gar_api_generator(paste0(base_url, ":runQuery"),
+                                             "POST",
+                                             data_parse_function = function(resp) resp)
+
+
+  cache_reset <- function() {
+    query_results <- query_ds(the_body = list(gqlQuery = list(queryString = paste0("SELECT * FROM ", cache_name))))
+    while((query_results$batch$moreResults != "NO_MORE_RESULTS") | is.null(query_results$batch$entityResults) == FALSE) {
+        ids <- (query_results$batch$entityResults$entity$key$path %>% dplyr::bind_rows())$name
```

Comment:
You will need to remove use of `%>%` and the dplyr dependency here.

I don't know exactly what data structure is being returned, but from the code this looks like you can use `vapply()` with an extractor function.